vms/evm/warp/backend.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package warp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/libevm/log"
+
+	"github.com/ava-labs/avalanchego/cache"
+	"github.com/ava-labs/avalanchego/cache/lru"
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p/acp118"
+	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
+	"github.com/ava-labs/avalanchego/vms/evm/uptimetracker"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp/payload"
+
+	avalancheWarp "github.com/ava-labs/avalanchego/vms/platformvm/warp"
+)
+
+var (
+	_                         Backend = (*backend)(nil)
+	ErrValidateBlock                  = errors.New("failed to validate block message")
+	ErrVerifyWarpMessage              = errors.New("failed to verify warp message")
+	errParsingOffChainMessage         = errors.New("failed to parse off-chain message")
+
+	messageCacheSize = 500
+)
+
+type BlockClient interface {
+	GetAcceptedBlock(ctx context.Context, blockID ids.ID) (snowman.Block, error)
+}
+
+// Backend tracks signature-eligible warp messages and provides an interface to fetch them.
+// The backend is also used to query for warp message signatures by the signature request handler.
+type Backend interface {
+	// AddMessage signs [unsignedMessage] and adds it to the warp backend database
+	AddMessage(unsignedMessage *avalancheWarp.UnsignedMessage) error
+
+	// GetMessageSignature validates the message and returns the signature of the requested message.
+	GetMessageSignature(ctx context.Context, message *avalancheWarp.UnsignedMessage) ([]byte, error)
+
+	// GetBlockSignature returns the signature of a hash payload containing blockID if it's the ID of an accepted block.
+	GetBlockSignature(ctx context.Context, blockID ids.ID) ([]byte, error)
+
+	// GetMessage retrieves the [unsignedMessage] from the warp backend database if available
+	GetMessage(messageHash ids.ID) (*avalancheWarp.UnsignedMessage, error)
+
+	acp118.Verifier
+}
+
+// backend implements Backend, keeps track of warp messages, and generates message signatures.
+type backend struct {
```

Comment:
I have a feeling @joshua-kim isn't going to like the complexity of this struct. My initial thought would to break it into 3 structs for each of the three tasks, but this adds another layer of indirection. can this be simplified somehow? 

Looking at it I have a feeling it is not that good, but I don't know enough to propose how to fix it properly.